pkg/scheduler/algorithm/priorities/resource_bin_packing.go

```diff
@@ -0,0 +1,240 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package priorities
+
+import (
+	"fmt"
+	"math"
+	"strings"
+
+	"github.com/golang/glog"
+	"k8s.io/api/core/v1"
+	v1helper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
+	"k8s.io/kubernetes/pkg/scheduler/algorithm"
+	priorityutil "k8s.io/kubernetes/pkg/scheduler/algorithm/priorities/util"
+	schedulerapi "k8s.io/kubernetes/pkg/scheduler/api"
+	schedulercache "k8s.io/kubernetes/pkg/scheduler/cache"
+)
+
+// ResourceBinPacking contains information to calculate bin packing priority.
+type ResourceBinPacking struct {
+	Resources []Resources
+}
+
+//Resources contains resource name and weight
+type Resources struct {
+	Resource v1.ResourceName
+	Weight   int
+}
+
+// NewResourceBinPacking creates a ResourceBinPackingPriorityMap.
+func NewResourceBinPacking(resources []Resources) (algorithm.PriorityMapFunction, algorithm.PriorityReduceFunction) {
+	for i, elm := range resources {
+		if elm.Weight == 0 {
+			resources[i].Weight = 1
+		}
+	}
+	resourceBinPackingPrioritizer := &ResourceBinPacking{
+		Resources: resources,
+	}
+	return resourceBinPackingPrioritizer.ResourceBinPackingPriorityMap, nil
+}
+
+// ResourceBinPackingPriorityDefault creates a resourceBinPacking based prioity
+func ResourceBinPackingPriorityDefault() *ResourceBinPacking {
+	defaultResourceBinPackingPrioritizer := &ResourceBinPacking{
+		Resources: []Resources{},
+	}
+	return defaultResourceBinPackingPrioritizer
+}
+
+// ResourceBinPackingPriorityMap is a priority function that favors nodes that have higher utlization of scare resource.
```

Comment:
nit: scare=scarce

Reply:
updated it to resource as people had issue with the term scarceResource